### PR TITLE
Moving Log-client stubs into separate file

### PR DESCRIPTION
**Pull request by Mercurrent**

### CODE REVIEW

File: submission/stub.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+func (m stubLogClient) AddPreChain(ctx context.Context, chain []ct.ASN1Cert) (*ct.SignedCertificateTimestamp, error) {
+	if _, ok := m.rootsCerts[m.logURL]; ok {
+		return testSCT(m.logURL), nil
+	}
+	return nil, fmt.Errorf("log %q has no roots", m.logURL)
+}
+
+func (m stubLogClient) GetAcceptedRoots(ctx context.Context) ([]ct.ASN1Cert, error) {
+	roots := []ct.ASN1Cert{}
+	if certInfos, ok := m.rootsCerts[m.logURL]; ok {
```
Comment (phad) [inline submission/stub.go]: If you change the check to !ok you can return early and then un-indent everything that follows.
(That also requires breaking the certInfos, ok assignment out of the if)

File: submission/stub.go [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+		for _, certInfo := range certInfos {
+			if len(certInfo.raw) > 0 {
+				roots = append(roots, ct.ASN1Cert{Data: certInfo.raw})
```
Comment (phad) [inline submission/stub.go]: If you follow this roots append with 'continue' you can remove the else and unindent the other if branch.

File: submission/stub.go [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+			} else {
+
+				roots = append(roots, ct.ASN1Cert{Data: readCertFile(certInfo.filename)})
+			}
+		}
+	}
+	return roots, nil
+}
+
+func buildRootedStubLc(log *loglist.Log, rCerts map[string][]rootInfo) (client.AddLogClient, error) {
```
Comment (phad) [inline submission/stub.go]: http://go/gh/golang/go/wiki/CodeReviewComments#initialisms
If you're going to shorten LogClient to Lc, it ought to be LC

File: submission/stub.go [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+	return stubLogClient{logURL: log.URL, rootsCerts: rCerts}, nil
+}
+
+func buildEmptyStubLogClient(log *loglist.Log) (client.AddLogClient, error) {
+	return buildRootedStubLc(log, map[string][]rootInfo{})
+}
+
+func buildStubLc(log *loglist.Log) (client.AddLogClient, error) {
```
Comment (phad) [inline submission/stub.go]: same here.

File: submission/stub.go [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+func (m stubLogClient) GetAcceptedRoots(ctx context.Context) ([]ct.ASN1Cert, error) {
+	roots := []ct.ASN1Cert{}
+	if certInfos, ok := m.rootsCerts[m.logURL]; ok {
+		for _, certInfo := range certInfos {
+			if len(certInfo.raw) > 0 {
+				roots = append(roots, ct.ASN1Cert{Data: certInfo.raw})
+			} else {
+
```
Comment (phad) [inline submission/stub.go]: remove empty line

File: submission/stub.go (hunk 3 above)
Comment (Mercurrent) [reply submission/stub.go]: Removed as unused by now.

File: submission/stub.go (hunk 4 above)
Comment (Mercurrent) [reply submission/stub.go]: Removed as unused by now.

File: submission/stub.go (hunk 5 above)
Comment (Mercurrent) [reply submission/stub.go]: Done.

File: submission/stub.go (hunk 2 above)
Comment (Mercurrent) [reply submission/stub.go]: Done.

File: submission/stub.go (hunk 1 above)
Comment (Mercurrent) [reply submission/stub.go]: Done.

**Pull request merged**